Introduce new script to automate the build process

ckelleyRH opened the pull request:
A new script `update_version.sh` is introduced, to start to reduce some of the manual burden of building packages. Usage is as:

`./update_version.sh <major> <minor> <update> <phase>`

...where `phase` is optional.

In this first iteration, it does the following:

* Updates the spec version to the new version provided
* Commits that change
* Creates a tag based on the new version provided
* Creates a source tarball based on the new version provided

Further updates to add resilience/further functionality will follow.

ckelleyRH commented:
@edewata - the major thing I'm not sure of is updating the `release_number` variable:
```
# Downstream release number:
# - development/stabilization (unsupported): 0.<n> where n >= 1
# - GA/update (supported): <n> where n >= 1
%global           release_number 0.1
```
Would you expect to change this variable through this script? I have never altered it myself, but it is clearly meant to be linked with the `phase` variable which I do update.

edewata commented:
@ckelleyRH I think so. We don't want to release incomplete code in the middle of development process into any Fedora versions because it could cause urgent problems for many people, so the build needs to be clearly marked with `0.<n>` release number and `alpha` phase. When we reach feature complete, we could release it in Rawhide with `0.<n>` release number and `beta` phase so people can try it but not use it in production. Once we reach code complete we could release it in the latest Fedora with the normal `<n>` release number (without any phase).

edewata commented:
I think the scrirpt generally looks good, but we probably should provide an instruction on how to undo the tag or revert/amend the commit in case we find a problem after running the script.

ckelleyRH commented:
> @ckelleyRH I think so. We don't want to release incomplete code in the middle of development process into any Fedora versions because it could cause urgent problems for many people, so the build needs to be clearly marked with `0.<n>` release number and `alpha` phase. When we reach feature complete, we could release it in Rawhide with `0.<n>` release number and `beta` phase so people can try it but not use it in production. Once we reach code complete we could release it in the latest Fedora with the normal `<n>` release number (without any phase).

OK, I'll have a think and do another iteration.

ckelleyRH commented:
> I think the scrirpt generally looks good, but we probably should provide an instruction on how to undo the tag or revert/amend the commit in case we find a problem after running the script.

I actually have a script to do this, I ran `autobuild.sh` so many times to test i got bored of typing `git reset; git tag -d`😅 

I will include a fixed up version of that, and a readme, in the next iteration.

ckelleyRH commented:
@edewata I addressed the following:

- Changing between alpha/beta and GA changes the `release_number` field.
- Renamed the script to better reflect what it does
- Added a check that the phase is a legit string
- Added some doc into the script
- Added a script that reverts the script, in case of :fire: 
- Updated the tests to check that the tag is created, if you get that far you've done OK.

Putting up for review again @fmarco76 @ladycfu @jmagne

edewata commented on .github/workflows/update-version-test.yml:
```diff
@@ -0,0 +1,41 @@
+
+      - name: Update to version with a phase
+        run: |
+          ./update_version.sh 12 1 1 beta1
```
To be more realistic, this should be `12.1.0-beta1`. We create alpha/beta releases before the GA which always ends with `.0`. After GA, we only release updates like `.1`, `.2`, etc. without any phases.

edewata commented on .github/workflows/update-version-test.yml:
```diff
@@ -0,0 +1,41 @@
+
+      - name: Update to version without a phase
+        run: |
+          ./update_version.sh 12 1 2
```
This can be changed to `12 1 0` to reflect the GA release.

edewata commented on .github/workflows/update-version-test.yml:
```diff
@@ -0,0 +1,41 @@
+
+      - name: Update to version with a phase again
+        run: |
+          ./update_version.sh 12 1 3 alpha1
```
We can use alpha/beta again in the next minor version before GA, i.e. `12.2.0-alpha1`.

edewata reviewed (commented):
Thanks for the update! I have some comments about the test. I'm still looking at the script.

ckelleyRH commented:
> Thanks for the update! I have some comments about the test. I'm still looking at the script.

Updated in next patchset, thanks!

ckelleyRH commented:
OK, latest revision incorporates @edewata's suggestions, fixes a typo with the phase in the version and fixes the script so it actually passes the test (yay tests!).

edewata commented on update_version.sh:
```diff
@@ -0,0 +1,100 @@
+            echo "Update phase to $NEXT_PHASE"
+            sed -i "/phase /c\%global           phase $NEXT_PHASE" pki.spec
+            echo "Update release_number"
+            sed -i "/release_number /c\%global           release_number 0.1" pki.spec
```
Ideally if we update the phase we should bump the release number too so that the whole NVR can be sorted properly, but this should work just fine with `alpha<N>` and `beta<N>` phases as long as `N` is a single digit, so I'll let you decide if you want to do that.

If you choose not to bump the release number, we might as well just use `0.` prefix instead of `0.1` if the phase exists.

ckelleyRH replied:
Hmm, good point - I'll take a look.

edewata replied:
Just FYI, ideally the phase should be part of the version number, and the release should be just `0.N`:
* https://docs.fedoraproject.org/en-US/packaging-guidelines/Versioning/#_complex_versioning
* https://docs.fedoraproject.org/en-US/packaging-guidelines/Versioning/#_prerelease_versions

No need to implement this now, I think it will require more testing, but maybe in the future we can properly follow this guideline.

ckelleyRH replied:
I updated the script so now `release number` is:

- No `phase` = 1 (first release)
- `phase` updated = 0.<$CURRENT_RELEASE_NUMBER++>
-  `phase` added = 0.1 (first pre-release)

We can make further modifications later if we decide to change the versioning.

ckelleyRH replied:
> I updated the script so now `release number` is:
> 
> * No `phase` = 1 (first release)
> * `phase` updated = 0.<$CURRENT_RELEASE_NUMBER++>
> * `phase` added = 0.1 (first pre-release)
> 
> We can make further modifications later if we decide to change the versioning.

Hi @edewata, I saw you were OK with this merging but before I do would like to check that this update to the `release_number` makes sense to you, if you would be so kind as to re-review :smile:. Then I will merge and backport etc. Thanks!

edewata replied:
Yup, LGTM.

edewata commented on update_version.sh:
```diff
@@ -0,0 +1,100 @@
+    CURRENT_UPDATE=$(grep "update_version " pki.spec | grep -Eo '[0-9]+$')
+    CURRENT_PHASE=$(grep "phase " pki.spec | grep -E 'alpha|beta' | awk '{print $(NF)}')
+
+    if [[ "$CURRENT_MAJOR" != "$NEXT_MAJOR" ]] ; then
+       echo "Update major version to $NEXT_MAJOR"
+       sed -i "/major_version /c\%global           major_version $NEXT_MAJOR" pki.spec
+    fi
+
+    if [[ "$CURRENT_MINOR" != "$NEXT_MINOR" ]] ; then
+        echo "Update minor version to $NEXT_MINOR"
+        sed -i "/minor_version /c\%global           minor_version $NEXT_MINOR" pki.spec
+    fi
+
+    if [[ "$CURRENT_UPDATE" != "$NEXT_UPDATE" ]] ; then
+        echo "Update update version to $NEXT_UPDATE"
+        sed -i "/update_version /c\%global           update_version $NEXT_UPDATE" pki.spec
+    fi
```
The conditions are probably not necessary. We can always update the version numbers regardless of the values, but I'll let you decide.

ckelleyRH replied:
I removed the conditions, thanks!

edewata reviewed (approved):
Thanks for the update! I have some comments below, but I think it LGTM. Feel free to update/merge.
